src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,669 @@
+		if (!IsResponseOK(result))
+		{
+			ReportResultError(connection, result, ERROR);
+		}
+
+		ClusterClock *nodeClockValue = ParseClusterClockPGresult(result, 0, 0);
+
+		ereport(DEBUG1, (errmsg("node(%lu:%u) transaction clock %lu:%u",
+								connection->connectionId, connection->port,
+								nodeClockValue->logical, nodeClockValue->counter)));
+
+		if (globalClockValue)
+		{
+			globalClockValue = LargerClock(globalClockValue, nodeClockValue);
+		}
+		else
+		{
+			globalClockValue = nodeClockValue;
+		}
+
+		PQclear(result);
+		ForgetResults(connection);
+	}
+
+	ereport(DEBUG1, (errmsg("final global transaction clock %lu:%u",
+							globalClockValue->logical,
+							globalClockValue->counter)));
+	return globalClockValue;
+}
+
+
+/*
+ * AdjustClocksToTransactionHighest Sets the clock value of all the nodes, participated
+ * in the PREPARE of the transaction, to the highest clock value of all the nodes.
+ */
+static void
+AdjustClocksToTransactionHighest(List *nodeConnectionList, ClusterClock
+								 *transactionClockValue)
+{
+	StringInfo queryToSend = makeStringInfo();
+
+	/* Set the adjusted value locally */
+	AdjustLocalClock(transactionClockValue);
+
+	/* Set the clock value on participating worker nodes */
+	appendStringInfo(queryToSend,
+					 "SELECT pg_catalog.citus_internal_adjust_local_clock_to_remote"
+					 "('(%lu, %u)'::pg_catalog.cluster_clock);",
+					 transactionClockValue->logical, transactionClockValue->counter);
+
+	ExecuteRemoteCommandInConnectionList(nodeConnectionList, queryToSend->data);
+}
+
+
+/*
+ * During prepare, once all the nodes acknowledge commit, persist the current
+ * transaction id along with the clock value in the catalog.
+ */
+ClusterClock *
+PrepareAndSetTransactionClock(void)
+{
+	if (!EnableClusterClock)
+	{
+		/* citus.enable_cluster_clock is false */
+		ereport(NOTICE, (errmsg("GUC enable_cluster_clock is off")));
+		return NULL;
+	}
+
+	if (!IsTransactionBlock())
+	{
+		ereport(NOTICE,
+				(errmsg("must be called inside a transaction and before a commit")));
+		return NULL;
+	}
+
+	dlist_iter iter;
+	List *transactionNodeList = NIL;
+
+	/* Prepare the connection list */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		Assert(transaction->transactionState != REMOTE_TRANS_NOT_STARTED);
+
+		/* Skip a transaction that failed */
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		/*
+		 * Check if any DML or DDL is executed over the connection on any
+		 * placement/table. If yes, add to the list, otherwise we skip
+		 * prepare since the connection didn't perform any write (read-only)
+		 */
+		if (ConnectionModifiedPlacement(connection))
+		{
+			transactionNodeList = lappend(transactionNodeList, connection);
+		}
+	}
+
+	/* include the transaction coordinator */
+	int connectionFlags = 0;
```

Comment:
two questions:
- what if the coordinator is already in `InProgressTransactions`? That could happen if the coordinator has placements
- What if this function is called from a worker node? Than, we'd have it twice

Reply:
BTW, coordinator, I meant the transaction coordinator, which can be any node in MX. But I will make that clear in the comment.

Reply:
>>what if the coordinator is already in InProgressTransactions? That could happen if the coordinator has placements

@onderkalaci I just checked again, InProgressTransactions doesn't include transaction-coordinator.

Reply:
>>What if this function is called from a worker node? Than, we'd have it twice

From worker too, we have only participating nodes in `InProgressTransactions`  and not the transaction-coordinator.

But I guess I need to verify if the local node *participated* in the transaction or not  before adding

Reply:
`InProgressTransactions`  is prepared from `StartRemoteTransactionBegin` that explain why current node is missing.